benchmark/src/DriverBench/ParallelBench.php

```diff
@@ -0,0 +1,199 @@
+use PhpBench\Attributes\ParamProviders;
+use PhpBench\Attributes\Revs;
+use RuntimeException;
+
+use function Amp\ParallelFunctions\parallelMap;
+use function Amp\Promise\wait;
+use function array_map;
+use function ceil;
+use function count;
+use function file;
+use function file_get_contents;
+use function file_put_contents;
+use function is_dir;
+use function mkdir;
+use function pcntl_fork;
+use function pcntl_waitpid;
+use function range;
+use function sprintf;
+use function str_repeat;
+use function sys_get_temp_dir;
+use function unlink;
+
+use const FILE_IGNORE_NEW_LINES;
+use const FILE_NO_DEFAULT_CONTEXT;
+use const FILE_SKIP_EMPTY_LINES;
+
+/**
+ * For accurate results, run benchmarks on a standalone server.
+ *
+ * @see https://github.com/mongodb/specifications/blob/ddfc8b583d49aaf8c4c19fa01255afb66b36b92e/source/benchmarking/benchmarking.rst#parallel
+ */
+#[BeforeClassMethods('beforeClass')]
+#[AfterClassMethods('afterClass')]
+final class ParallelBench
+{
+    /** @var string[] */
+    private static array $files = [];
+
+    public static function beforeClass(): void
+    {
+        // Generate files
+        $fileContents = str_repeat(file_get_contents(Data::LDJSON_FILE_PATH), 5_000);
+        foreach (self::getFileNames() as $file) {
+            file_put_contents($file, $fileContents);
+        }
+    }
+
+    public static function afterClass(): void
+    {
+        foreach (self::getFileNames() as $file) {
+            unlink($file);
+        }
+
+        self::$files = [];
+    }
+
+    /**
+     * Parallel: LDJSON multi-file import
+     * Using single thread
+     */
+    #[BeforeMethods('beforeMultiFileImport')]
+    #[Revs(1)]
+    public function benchMultiFileImport(): void
+    {
+        $collection = Utils::getCollection();
+        foreach (self::getFileNames() as $file) {
+            self::importFile($file, $collection);
+        }
+    }
+
+    /**
+     * Parallel: LDJSON multi-file import
+     * Using multiple forked threads
+     *
+     * @param array{processes:int, files:string[], batchSize:int} $params
+     */
+    #[BeforeMethods('beforeMultiFileImport')]
+    #[ParamProviders(['provideProcessesParameter'])]
+    #[Revs(1)]
+    public function benchMultiFileImportFork(array $params): void
+    {
+        $pids = [];
+        foreach (self::getFileNames() as $file) {
+            // Wait for a child process to finish if we have reached the maximum number of processes
+            if (count($pids) >= $params['processes']) {
+                $pid = pcntl_waitpid(-1, $status);
+                unset($pids[$pid]);
+            }
+
+            $pid = pcntl_fork();
+            if ($pid === 0) {
+                // If we reset, we can garantee that we get a new manager in the child process
+                // If we don't reset, we will get the same manager client_zval in the child process
+                // and share the libmongoc client.
+                Utils::reset();
```

Comment:
Comment dictated by @jmikola.
Unsetting the Client/Database/Collection objects is required.

Reply:
I suggest:

> Reset to ensure that the existing libmongoc client (via the Manager) is not re-used by the child process. When the child process constructs a new Manager, the differing PID will result in creation of a new libmongoc client.

Two additional thoughts:

You can consider handling this in `Utils::getClient()` directly by having it track the PID used to create the `$client` singleton. Then, if the PID ever changes, `getClient()` can reset on its own. This might require eliminating the `$database` and `$collection` singletons, though (or applying the PID logic to those getters).

Additionally, I think there's probably no harm in using `disableClientPersistence: true` when creating clients if you don't actually expect to create multiple instances. This wouldn't impact forking at all, though, so I wouldn't say it's necessary.

Reply:
I think it's better staying explicit in this benchmark: it teaches the best practice to potential readers and it doesn't impact other benchmarks that don't uses forks.

Reply:
Comment updated.

Reply:
> and it doesn't impact other benchmarks that don't uses forks.

Ah, if this is the only file that will utilize forking, then I agree with keeping the logic here.